packages/NodeTypeResolver/NodeTypeResolver.php

```diff
@@ -438,7 +438,7 @@ private function isMatchingUnionType(Type $resolvedType, ObjectType $requiredObj
             return $this->isMatchObjectWithoutClassType($type, $requiredObjectType);
         }
 
-        return $type->isSuperTypeOf($requiredObjectType)
+        return $requiredObjectType->isSuperTypeOf($type)
             ->yes();
```

Comment:
@TomasVotruba here is the fix, the `->superTypeOf()` must use `$requiredObjectType` like used in other method for object type check:

https://github.com/rectorphp/rector-src/blob/2ad03db37f187d715c3a2f189dc181063126e596/packages/NodeTypeResolver/NodeTypeResolver.php#L405-L411